### PR TITLE
bbb-record: refactoring, more config checks, list-workflows option

### PR DESCRIPTION
**Pull request by danielpetri1**
### What does this PR do?
Extends BigBlueButton's recording diagnostic utility with permission and ownership checks that may help in circumstances where recording processing is not starting.

Checks were written based on the configuration set in the `after-install.sh` script, community feedback as in [#12627](https://github.com/bigbluebutton/bigbluebutton/issues/12627) and [#9279 ](https://github.com/bigbluebutton/bigbluebutton/issues/9279), and personal experience.

- The`list-recent` option has been documented in `--help`.
- Introduces a `--list-workflows` option.
- Informs about playback protocol.

### Motivation
There were no configuration checks for the default presentation format.

Developed for the Open Source Lab at TUM.

example output

### TODO
[x] Lint with shellcheck:
- Replace cat with `'cmd < file | ..' or 'cmd file | ..'`
- Use `./*glob*` or `-- *glob*` so names with dashes won't become options
- Use `"${var:?}"` to ensure `rm -rf` never expands to `/` .
- Prefer `[ p ] && [ q ]` as `[ p -a q ]` is not well defined.
- Use `${variable//search/replace}` instead of `$(echo "$variable" | sed 's/search/replace/g')`
- Substitute deprecated `egrep` with `grep -E`
- Use `..| sudo tee file` in redirects
- Double quote to prevent globbing and word splitting

### CODE REVIEW

**Comment by antobinary** [comment]
Does this PR resolve https://github.com/bigbluebutton/bigbluebutton/issues/9279 ?

**Comment by danielpetri1** [comment]
Hi @antobinary,
While the new tests would have pointed out the wrong ownership in #9279, no automatic fix would have been attempted.
In the latest commit one more check for the raw directory has been added.

Maybe we could consider checks for the playback directory as well, as in [#14026](https://github.com/bigbluebutton/bigbluebutton/issues/14026)? I haven't touched that for now since it's a separate module, but maybe it's worth discussing it.

**Comment by kepstin** [comment]
This is a really hard PR to review, because you're changing nearly the entire file, including parts unrelated to the subject of the PR. (I guess this is partly due to applying shellcheck fixes throughout the script?)

I think I agree that these changes should happen, but I'm not sure about whether doing them all in a single PR makes sense. At the very least, the subject of the PR should be expanded to say more than just "Additional bbb-record checks"

File: bigbluebutton-config/bin/bbb-record [hunk 1]
```diff
@@ -45,20 +45,31 @@
...
-RAW_PRESENTATION_SRC=$(cat /usr/local/bigbluebutton/core/scripts/bigbluebutton.yml | sed -n '/\(raw_presentation_src\)/{s/.*raw_presentation_src:[ ]*//;s/;//;p}')
+BBB_PROPERTIES=/usr/share/bbb-web/WEB-INF/classes/bigbluebutton.properties
+BBB_YML=/usr/local/bigbluebutton/core/scripts/bigbluebutton.yml
+
+SERVER_URL=$(sed -n '/^bigbluebutton.web.serverURL/{s/.*\///;p}' < $BBB_PROPERTIES)
+
+# $1: Desired property
+# Only returns 'PRODUCTION' values
+get_yml_value(){
+   sed -n '/\('"$1"'\)/{s/.*'"$1"':[ ]*//;s/;//;p}' < $BBB_YML | head -1
```
Comment (kepstin) [inline bigbluebutton-config/bin/bbb-record:60]: We require that the `yq` tool is installed already for other bigbluebutton scripts, we should probably use that here as well.

The get_yml_value() function could just be:
`yq read "$BBB_YML" "$1"`

File: bigbluebutton-config/bin/bbb-record [hunk 2]
```diff
@@ -48,10 +48,10 @@
 source /etc/bigbluebutton/bigbluebutton-release
 
 BBB_USER=bigbluebutton
-BBB_PROPERTIES=/usr/share/bbb-web/WEB-INF/classes/bigbluebutton.properties
 BBB_YML=/usr/local/bigbluebutton/core/scripts/bigbluebutton.yml
 
-SERVER_URL=$(sed -n '/^bigbluebutton.web.serverURL/{s/.*\///;p}' < $BBB_PROPERTIES)
+# BBB_PROPERTIES=/usr/share/bbb-web/WEB-INF/classes/bigbluebutton.properties
```
Comment (kepstin) [inline bigbluebutton-config/bin/bbb-record]: Please don't leave commented code hanging around, at least without a comment explaining why someone might want to uncomment and use this code on a live BBB system.

**Comment by danielpetri1** [comment]
Hi @kepstin, 
Thanks for taking a look at it. I tried to keep new functionality changes separate from the refactoring/linting ones through individual commits. I do understand that that may not have been the best approach. Do you want me to create two separate PRs and cherry-pick the changes?

File: bigbluebutton-config/bin/bbb-record (hunk 2 above)
Comment (danielpetri1) [reply bigbluebutton-config/bin/bbb-record]: Should I just remove it, then? This variable was previously declared but not used anywhere.

File: bigbluebutton-config/bin/bbb-record (hunk 1 above)
Comment (danielpetri1) [reply bigbluebutton-config/bin/bbb-record:60]: Good idea, was unaware of that tool.

**Comment by kepstin** [comment]
I think it's probably OK to have as all one PR, just please edit the PR message to explain everything that the PR is doing so the message when it's merged is complete.

File: bigbluebutton-config/bin/bbb-record (hunk 2 above)
Comment (kepstin) [reply bigbluebutton-config/bin/bbb-record]: Yeah, if the code is unused and you can't think of a reason someone would uncomment it on a running system, then just remove it.

File: bigbluebutton-config/bin/bbb-record [hunk 3]
```diff
@@ -737,6 +747,17 @@ fi
 #   fi
 #fi
 
+if [ $LISTWORKFLOWS ]; then
+   echo
+   echo "# Available workflows:"
+   echo "#"
+   for type in $TYPES; do
```
[the remark this answers is not in the document]
Comment (danielpetri1) [reply bigbluebutton-config/bin/bbb-record:785]: Right. So my idea was to list the formats that can be turned on with the `--enable <workflow>` option (making it easier to tell whether the screenshare and podcast formats are installed for instance). I've changed it now to list the current recording steps, as well as the available processing scripts. Is this sensible? Here's an example output.

```bash
# Enabled recording processing steps:
#
# archive
# sanity
# captions
# process:presentation
# publish:presentation
#
# Available processing scripts: 
#
# breakout_pdf
# presentation
```

[the remark this answers is not in the document]
Comment (kepstin) [reply bigbluebutton-config/bin/bbb-record:785]: I like this, it provides a helpful overview for folks trying to debug why their recordings aren't being processed in a specific format.

**Review by kepstin — APPROVED** [review]

**Comment by antobinary** [comment]
Hi @danielpetri1 
Could you please include a [draft?] PR for documenting these changes on
https://docs.bigbluebutton.org/dev/recording.html#manage-recordings 

https://github.com/bigbluebutton/bigbluebutton.github.io

If possible, please express in the docs that the new option(s) are introduced in BBB 2.5

**Pull request merged**